server/conf/evolutions/default/44.sql

```diff
@@ -0,0 +1,66 @@
+# --- Add database triggers to set the latest_status column on creating an application_event or
+# --- updating an application.
+# --- For rationale behind choosing database triggers, see:
+# ---   https://github.com/civiform/civiform/issues/3269#issuecomment-1232193163
+# --- Relevant documentation:
+# --- Trigger functions: https://www.postgresql.org/docs/current/sql-createtrigger.html
+# --- Create trigger statement: https://www.postgresql.org/docs/current/sql-createtrigger.html
+
+# --- !Ups
+
+alter table applications add column if not exists latest_status varchar;
+
+-- The latest status associated with an application is set to NULL if no application event entries
+-- exist. Otherwise, the text from the latest event corresponding to a status change is used.
+CREATE OR REPLACE FUNCTION retrieve_latest_status(app_id applications.id%TYPE) RETURNS text AS $$
+  SELECT
+      (CASE WHEN details->'status_event'->>'status_text' = ''
+        THEN NULL
+        ELSE details->'status_event'->>'status_text'
+       END) AS latest_status
+      FROM application_events
+      WHERE
+        application_id = app_id AND
+        details->>'event_type' = 'STATUS_CHANGE'
+      ORDER BY create_time DESC
+      LIMIT 1
+$$ LANGUAGE SQL;
+
+-- A function that sets the "latest_status" column anytime an application is updated.
+-- Note: The "NEW" variable corresponds to the application record to be inserted / updated.
+CREATE OR REPLACE FUNCTION process_application_change() RETURNS TRIGGER AS $$
+  BEGIN
+    NEW.latest_status := retrieve_latest_status(NEW.id);;
+    RETURN NEW;;
+  END;;
+$$ LANGUAGE plpgsql;
+
+CREATE TRIGGER application_change BEFORE INSERT OR UPDATE ON applications
+    FOR EACH ROW EXECUTE FUNCTION process_application_change();
+
+-- A function that sets the "latest_status" column on the referenced application record anytime
+-- an event record is updated.
+-- Note: The "NEW" variable corresponds to the application event record to be inserted / updated.
+CREATE OR REPLACE FUNCTION process_application_event_change() RETURNS TRIGGER AS $$
+  BEGIN
+    IF ((NEW.details->>'event_type') = 'STATUS_CHANGE') THEN
+      UPDATE applications
+        SET latest_status = retrieve_latest_status(NEW.application_id)
```

Comment:
This function is done ON the Events table so the right side here is unnecessarily re-querying the Events table I think; can't it be something like

NEW.details->'status_event'->>'status_text'

Reply:
Technically yes. However, say we were modifying an old event (either through a DB migration or through new functionality we add). In that case, we still want to set the latest status based on the most recent event. Given that, the query is necessary. Added docs that attempt to clarify this.